spec/unit/shoulda/matchers/active_record/have_attached_matcher_spec.rb

```diff
@@ -0,0 +1,284 @@
+          end
+          it 'matches' do
+            expect { have_many_attached(:avatars) }.
+              not_to match_against(record).
+              and_fail_with(<<-MESSAGE)
+Expected User to have a has_many_attached called avatars, but this could not be proved.
+  User does not have a :avatars= method.
+            MESSAGE
+          end
+        end
+        context 'and the attachments association does not exist' do
+          let(:record) do
+            record_has_many_attached(:avatars, remove_attachments: true)
+          end
+          it 'matches' do
+            expect { have_many_attached(:avatars) }.
+              not_to match_against(record).
+              and_fail_with(<<-MESSAGE)
+Expected User to have a has_many_attached called avatars, but this could not be proved.
+  Expected User to have a has_many association called avatars_attachments (no association called avatars_attachments)
+            MESSAGE
+          end
+        end
+        context 'and the blobs association is invalid' do
+          let(:record) do
+            record_has_many_attached(:avatars, invalidate_blobs: true)
+          end
+          it 'matches' do
+            expect { have_many_attached(:avatars) }.
+              not_to match_against(record).
+              and_fail_with(<<-MESSAGE)
+Expected User to have a has_many_attached called avatars, but this could not be proved.
+  Expected User to have a has_many association called avatars_blobs (avatars_blobs should resolve to ActiveStorage::Blob for class_name)
+            MESSAGE
+          end
+        end
+
+        context 'and the eager loading scope does not exist' do
+          let(:record) do
+            record_has_many_attached(:avatars, remove_eager_loading_scope: true)
+          end
+          it 'matches' do
+            expect { have_many_attached(:avatars) }.
+              not_to match_against(record).
+              and_fail_with(<<-MESSAGE)
+Expected User to have a has_many_attached called avatars, but this could not be proved.
+  User does not have a :with_attached_avatars scope.
+            MESSAGE
+          end
+        end
+      end
+    end
+  end
+end
+
+def record_has_one_attached(
+  attached_name,
+  model_name: 'User',
+  remove_reader: false,
+  remove_writer: false,
+  remove_attachments: false,
+  invalidate_blobs: false,
+  remove_eager_loading_scope: false
+)
+  model = define_model(model_name) do
+    has_one_attached attached_name
+
+    if remove_reader
+      undef_method attached_name
+    end
+
+    if remove_writer
+      undef_method "#{attached_name}="
+    end
+
+    if remove_attachments
+      reflections.delete("#{attached_name}_attachment")
+    end
+
+    if invalidate_blobs
+      reflections["#{attached_name}_blob"].options[:class_name] = 'User'
+    end
+
+    if remove_eager_loading_scope
+      instance_eval <<-CODE, __FILE__, __LINE__ + 1
+undef with_attached_#{attached_name}
```

Comment:
Does it not work to use `undef_method` here?

Reply:
Well in some way this was not working. The `with_attached_xxx` is a rails scope scope defined by the `has_
one_attached` method. In some way it's not actually defining a class method that could easily be removed. If I try the same with a method that is explicitly defined, it does work. So this is still the best I could come up with for now.

I again tried it now and it still doesn't work 😞

Reply:
Hmm... okay! No problem